bootlib/bootlib.c

```diff
@@ -0,0 +1,80 @@
+//
+// Created by iliaz on 9/10/2024.
+//
+
+#include "bootlib.h"
+
+/**
+ * resets the boot counter.
+ * should be called after nominal operation is observed
+ */
+void reset_boot_counter()
+{
+    uint32_t variables=(uint32_t)(&__variables_start__);// "&" does not dereference a pointer. The value of __variables_start__ is put to the data pointer
+    uint32_t *data = (uint32_t *) variables;
+    uint32_t data_buffer[IFLASH_PAGE_SIZE/sizeof(uint32_t)];
+
+    data_buffer[BOOT_COUNTER]=0;//reset counter
+    data_buffer[PRIMARY_PARTITION_VARIABLE]=data[PRIMARY_PARTITION_VARIABLE];//keep primary firmware the same
+    data_buffer[BOOT_STATUS_VARIABLE]=data[BOOT_STATUS_VARIABLE];//keep boot status variable the same
+
+    for(uint32_t i=3; i<IFLASH_PAGE_SIZE/sizeof(uint32_t); i++) // reduce memory wear
+    {
+        data_buffer[i]=0xFFFFFFFF;
+    }
+
+    EFC_SectorErase(variables);
+    EFC_PageWrite(data_buffer,variables);
+}
+
+
+/**
+ * returns the ID of the primary firmware.
+ * (0->FIRST_FIRMWARE_IS_PRIMARY
+ * 1->SECOND_FIRMWARE_IS_PRIMARY)
+ */
+uint32_t get_primary_firmware()
+{
+    uint32_t variables=(uint32_t)(&__variables_start__);// "&" does not dereference a pointer. The value of __variables_start__ is put to the data pointer
+    uint32_t *data = (uint32_t *) variables;
```

Comment:
Why not just...
```suggestion
    uint32_t *data = (uint32_t *) (&__variables_start__);
```

Also, try defining this globally, I don't see a need to redefine it in every function. And make sure it's `volatile static`.

Reply:
Also, please don't call things `data`, make the name more specific.